File: app/src/main/java/com/xephorium/crystalnote/ui/custom/NoteToolbar.java
package com.xephorium.crystalnote.ui.custom;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;

import com.xephorium.crystalnote.R;

public class NoteToolbar extends Toolbar {

    private static final int DEFAULT_LEFT_BUTTON_IMAGE = R.drawable.icon_back;
    private static final int DEFAULT_RIGHT_BUTTON_IMAGE = R.drawable.icon_save;
    public static final int NO_IMAGE = -1;

    private boolean editMode;
    private NoteToolbarListener noteToolbarListener;

    public NoteToolbar(Context context) {
        super(context);
        buildNoteToolbar(context);
    }

    public NoteToolbar(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        buildNoteToolbar(context);
    }

    public NoteToolbar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        buildNoteToolbar(context);
    }

    private void buildNoteToolbar(Context context) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        View noteToolbarLayout = layoutInflater.inflate(R.layout.note_toolbar_layout, null);
        this.addView(noteToolbarLayout);

        this.setEditMode(false);
        this.setLeftButtonImage(DEFAULT_LEFT_BUTTON_IMAGE);
        this.setRightButtonImage(NO_IMAGE);
        this.noteToolbarListener = getDefaultNoteToolbarListener();

        findViewById(R.id.left_button).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                noteToolbarListener.onLeftButtonClick();
            }
        });
        findViewById(R.id.right_button).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                noteToolbarListener.onRightButtonClick();
            }
        });
        ((EditText) findViewById(R.id.edit_text)).addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                noteToolbarListener.onTextChange(
                        ((EditText) findViewById(R.id.edit_text)).getText().toString());
            }
        });
    }

    @Override
    public void setTitle(int stringResource) {
        getTitleView().setText(getResources().getText(stringResource));
    }

    @Override
    public CharSequence getTitle() {
        return getTitleView().getText();
    }

    public void setTitleContent(String content) {
        getEditView().setText(content);
    }

    public String getTitleContent() {
        return getEditView().getText() != null ? getEditView().getText().toString() : "";
    }

    public void setEditMode(boolean editMode) {
        if (editMode) {
            getEditView().setVisibility(VISIBLE);
            getTitleView().setVisibility(GONE);
        } else {
            getEditView().setVisibility(GONE);
            getTitleView().setVisibility(VISIBLE);
        }
    }

    public boolean isEditMode() {
        return this.editMode;
    }

    public void setLeftButtonImage(int drawable) {
        if (drawable == NO_IMAGE) {
            getLeftButtonView().setVisibility(GONE);
        } else {
            getLeftButtonView().setVisibility(VISIBLE);
            getLeftButtonView().setImageDrawable(getResources().getDrawable(drawable, getContext().getTheme()));
        }
    }

    public void setRightButtonImage(int drawable) {
        if (drawable == NO_IMAGE) {
            getRightButtonView().setVisibility(GONE);
        } else {
            getRightButtonView().setVisibility(VISIBLE);
            getRightButtonView().setImageDrawable(getResources().getDrawable(drawable, getContext().getTheme()));
        }
    }

    public void setNoteToolbarListener(NoteToolbarListener noteToolbarListener) {
        this.noteToolbarListener = noteToolbarListener;
    }

    private ImageView getLeftButtonView() {
        return ((ImageView) ((RelativeLayout) getChildAt(0)).getChildAt(0));
    }

    private ImageView getRightButtonView() {
        return ((ImageView) ((RelativeLayout) getChildAt(0)).getChildAt(1));
    }

    private TextView getTitleView() {
        return ((TextView) ((RelativeLayout) getChildAt(0)).getChildAt(3));
    }

    private EditText getEditView() {
        return ((EditText) ((RelativeLayout) getChildAt(0)).getChildAt(2));
    }

    private NoteToolbarListener getDefaultNoteToolbarListener() {
        return new NoteToolbarListener() {
            @Override
            public void onLeftButtonClick() {
            }

            @Override
            public void onRightButtonClick() {
            }

            @Override
            public void onTextChange(String text) {
            }
        };
    }

    public interface NoteToolbarListener {
        void onLeftButtonClick();

        void onRightButtonClick();

        void onTextChange(String text);
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/custom/NoteToolbar.kt
package com.xephorium.crystalnote.ui.custom

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.widget.Toolbar

import com.xephorium.crystalnote.R
import com.xephorium.crystalnote.data.model.CrystalNoteTheme
import com.xephorium.crystalnote.ui.extensions.getThemeColor

class NoteToolbar : Toolbar {


    /*--- Variable Declarations ---*/

    private var noteToolbarListener: NoteToolbarListener? = null

    var isEditMode: Boolean = false
        set(editMode) = if (editMode) {
            findViewById<EditText>(R.id.textToolbarEdit)?.visibility = View.VISIBLE
            findViewById<TextView>(R.id.textToolbarTitle)?.visibility = View.GONE
        } else {
            findViewById<EditText>(R.id.textToolbarEdit)?.visibility = View.GONE
            findViewById<TextView>(R.id.textToolbarTitle)?.visibility = View.VISIBLE
        }


    /*--- Toolbar Setup ---*/

    constructor(context: Context) : super(context) {
        buildNoteToolbar(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        buildNoteToolbar(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        buildNoteToolbar(context)
    }

    private fun buildNoteToolbar(context: Context) {
        val layoutInflater = LayoutInflater.from(context)
        val noteToolbarLayout = layoutInflater.inflate(R.layout.note_toolbar_layout, null)
        this.addView(noteToolbarLayout)

        this.isEditMode = false
        this.setLeftButtonImage(DEFAULT_LEFT_BUTTON_IMAGE)
        this.noteToolbarListener = getDefaultNoteToolbarListener()

        findViewById<ColorOrb>(R.id.colorOrbToolbar).run {
            setPadding(R.dimen.toolbarColorOrbPadding)
            setBackdropColor(CrystalNoteTheme.fromCurrentTheme(context).colorToolbar)
            setOutlineColor(CrystalNoteTheme.fromCurrentTheme(context).colorToolbarTextSecondary)
            setOutlineAlpha(COLOR_ORB_ALPHA)
            setOnClickListener { noteToolbarListener?.onColorClick() }
        }

        findViewById<ImageView>(R.id.buttonToolbarLeft).setOnClickListener {
            noteToolbarListener?.onLeftButtonClick()
        }

        findViewById<ImageView>(R.id.buttonToolbarRight).setOnClickListener {
            noteToolbarListener?.onRightButtonClick()
        }

        findViewById<EditText>(R.id.textToolbarEdit).addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                noteToolbarListener?.onTextChange(
                    (findViewById<View>(R.id.textToolbarEdit) as EditText).text.toString())
            }
        })
    }


    /*--- Public Methods ---*/

    override fun setTitle(stringResource: Int) {
        findViewById<TextView>(R.id.textToolbarTitle).text = resources.getText(stringResource)
    }

    fun setTitle(string: String) {
        findViewById<TextView>(R.id.textToolbarTitle).text = string
    }

    override fun getTitle(): CharSequence {
        return findViewById<TextView>(R.id.textToolbarTitle).text.toString()
    }

    fun setEditTextContent(content: String) {
        findViewById<EditText>(R.id.textToolbarEdit).setText(content)
    }

    fun getEditTextContent(): String {
        return findViewById<EditText>(R.id.textToolbarEdit).text.toString()
    }

    fun setLeftButtonImage(drawable: Int) {
        findViewById<ImageView>(R.id.buttonToolbarLeft).run {
            if (drawable == NO_IMAGE) {
                visibility = View.GONE
            } else {
                visibility = View.VISIBLE
                setImageDrawable(resources.getDrawable(drawable, context.theme))
                setColorFilter(
                    context.getThemeColor(R.attr.themeToolbarTextSecondary),
                    android.graphics.PorterDuff.Mode.SRC_IN
                )
                setLeftButtonScale(TOOLBAR_ICON_SCALE_SMALLER)
            }
        }
    }

    fun setLeftButtonScale(scale: Float) {
        findViewById<ImageView>(R.id.buttonToolbarLeft).run {
            scaleX = scale
            scaleY = scale
        }
    }

    fun showColor() {
        findViewById<ColorOrb>(R.id.colorOrbToolbar).visibility = View.VISIBLE
        findViewById<ColorOrb>(R.id.colorOrbToolbar).forceThickOutline()
    }

    fun hideColor() {
        findViewById<ColorOrb>(R.id.colorOrbToolbar).visibility = View.GONE
    }

    fun setColor(color: Int) {
        findViewById<ColorOrb>(R.id.colorOrbToolbar).setColor(color)
    }

    fun showRightButton() {
        findViewById<ImageView>(R.id.buttonToolbarRight).visibility = View.VISIBLE
    }

    fun hideRightButton() {
        findViewById<ImageView>(R.id.buttonToolbarRight).visibility = View.GONE
    }

    fun setNoteToolbarListener(noteToolbarListener: NoteToolbarListener) {
        this.noteToolbarListener = noteToolbarListener
    }


    /*--- Private Utility Methods ---*/

    private fun getDefaultNoteToolbarListener() = object : NoteToolbarListener {
        override fun onLeftButtonClick() {}
        override fun onRightButtonClick() {}
        override fun onColorClick() {}
        override fun onTextChange(text: String) {}
    }


    /*--- Action Handling Interface ---*/

    interface NoteToolbarListener {
        fun onLeftButtonClick()
        fun onRightButtonClick()
        fun onColorClick()
        fun onTextChange(text: String)
    }


    /*--- Constants ---*/

    companion object {
        private const val DEFAULT_LEFT_BUTTON_IMAGE = R.drawable.icon_back
        private const val COLOR_ORB_ALPHA = 0.68
        const val NO_IMAGE = -1
        const val TOOLBAR_ICON_SCALE_SMALLER = (0.93).toFloat()
    }
}
